### PR TITLE
Adds new extension point that permits appending to the top or bottom of a given widget

### PR DESCRIPTION
**Pull request by carolinebridge-oicr**
> Adds new extension point that permits appending to the top or bottom of a given widget

Just an idea I had when drafting possible ways to extend the bookmarks plugin via an external plugin. Pretty simple and similar to the existing `Core-replaceWidget` without completely replacing the existing UI; I can see plugin developers using this extension point.

### CODE REVIEW

File: website/docs/developer_guides/extension_points.md [hunk 1]
```diff
@@ -356,7 +356,7 @@ Example: replaces about feature details widget for a particular track ID
 
 ```typescript
 pluginManager.addToExtensionPoint(
-  'Core-replaceAbout',
+  'Core-replaceWidget',
```
Comment (cmdcolin) [inline website/docs/developer_guides/extension_points.md:359]: this appears to remove replaceAbout, but replaceAbout is still a thing that exists :)

Comment (carolinebridge-oicr) [reply website/docs/developer_guides/extension_points.md:359]: `Core-replaceAbout` has its own docs on line 266 of this file, i think its docs were replicated in the `Core-replaceWidget` section by accident

Comment (cmdcolin) [reply website/docs/developer_guides/extension_points.md:359]: nice catch

**Comment by cmdcolin** [comment]
one thing that may be notable is that the 'bottom' configuration goes after a 300px padding block for the drawer widget. might want to put it above the 300px padding block

**Comment by rbuels** [comment]
Technically, I am pretty sure anything you could do with `Core-addToWidget` you could already do with `Core-replaceWidget`, because you could just wrap the widget in a fragment and return `<><OriginalWidget /><MyAdditionalStuff /><>`.

But, that is probably not the easiest thing for third-party developers to understand, is it? What do y'all think?

**Comment by carolinebridge-oicr** [comment]
@rbuels I was trying to do that originally, but ran into issues with utilizing jsx in the `index.ts` file where the extension point is utilized, and further with retrieving the <OriginalComponent /> and its model outside of the extension point callback.

If you have a pattern that works and is simplistic enough to satisfy what this PR satisfies, we could consider adding it to the docs instead I suppose.

**Comment by rbuels** [comment]
Yeah this is probably better.

**Comment by cmdcolin** [comment]
I do like the idea of what @rbuels suggested since it is embracing that 'compositional' approach

example

```
pluginManager.addToExtensionPoint(
  "Core-replaceWidget",
  (DefaultWidget, { model }) => {
    return function NewWidget(props) {
      return (
        <div>
          <div>Hello world</div>
          <DefaultWidget {...props} />
        </div>
      );
    };
  },
);
```

it captures DefaultWidget via "closure" there, and returns a new component

**Comment by rbuels** [comment]
> I was trying to do that originally, but ran into issues with utilizing jsx in the index.ts file where the extension point is utilized, and further with retrieving the and its model outside of the extension point callback.

@carolinebridge-oicr are those issues surmountable? if you renamed index.ts to index.tsx, could you use jsx in the extension point? And does the existing extension point get the right arguments to be able to extend the component like that?

**Comment by cmdcolin** [comment]
then plugin-development-tools currently hardcodes the input file as src/index.ts, but this may be something we want to avoid

https://github.com/GMOD/jbrowse-components/blob/main/plugin-development-tools/src/createRollupConfig.ts#L117


temp workaround you can make inside the plugin template's rollup.config.js to fix, may be we want to allow either or just make src/index.tsx the default

```

export default createRollupConfig(globals.default, {
  includeUMD,
  includeCJS,
  includeESMBundle,
  includeNPM,
}).map((r) => ({ ...r, input: 'src/index.tsx' }))
```

**Comment by cmdcolin** [comment]
I added an example of using the 'compositional approach' to adding the component to the docs here https://github.com/GMOD/jbrowse-components/commit/c85a4bc1c1098eb2a01ae9e60242c9a2690fc4de

**Comment by cmdcolin** [comment]
possible arguments to keep this PR include

- ease of use
- force prepend even when another component in the replaceWidget chain does not obey the composition


i think for 1, the documented workaround is not bad, so may not be needed
i think for 2, i think this should basically be something we dont try to add workarounds for. plugins can obey composition or do total overrides if they want

**Comment by cmdcolin** [comment]
with my above assessment, could maybe close...let me know if there are use cases that need this though

**Pull request closed without merging**